scripts/cluster-upgrade.sh

```diff
@@ -255,45 +262,90 @@ upgrade() {
 
+        CO_COUNT_TO=$(echo "$CO_JSON" | jq -r '.items[].status.versions[] | select(.name == "operator") | .version' | grep $TO | wc -l)
+
+        # do not print status if there's nothing to print (i.e. the get failed for some reason)        
+        if [ $CO_COUNT -gt 0 ];
+        then
+            log $OCM_NAME "upgrade" "ClusterOperators: Progressing: $PROGRESSING_COUNT, Degraded: $DEGRADED_COUNT, Upgraded: $CO_COUNT_TO/$CO_COUNT"
+        fi
+
+        # check status if we are not at the same version
         check_cluster_status $OCM_NAME $CD_NAMESPACE $CD_NAME "upgrade"
+
+        # wait a (relatively) short time before checking version (and status when we look back through, if we do)
+        sleep 60
+        OCP_CURRENT_VERSION=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get clusterversion version -o json | jq -r '.status.history[] | select(.verified == true and .state == "Completed") | .version' | grep $TO)
     done
 
     log $OCM_NAME "upgrade" "ClusterVersion on $TO"
-    log $OCM_NAME "upgrade" "checking kubelet versions"
+    log $OCM_NAME "upgrade" "checking node versions"
 
     # fun fact!  after clusterversion says it is done, individual nodes could still be updated.
+    MASTER_KUBELET_VERSION=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get nodes --no-headers | grep master | sort | uniq | awk '{print $5}' | sort | uniq)
+
+    MACHINE_COUNT_ALL=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get nodes --no-headers | wc -l)
+
+    # could do joins from machine to node, but it's just all non-master at this time
+    MACHINE_COUNT_UPGRADED=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get nodes --no-headers | grep $MASTER_KUBELET_VERSION | wc -l)
+
+    MACHINE_COUNT_PENDING=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get nodes --no-headers | grep -v $MASTER_KUBELET_VERSION | wc -l)
+
+    # start maintenance only if we need to
+    if [ ! $MACHINE_COUNT_UPGRADED -ge $MACHINE_COUNT_ALL ];
+    then
+        # Create a new maintenance window for the kubelet (non master) upgrades based on the number of nodes.
+        MAINT_WINDOW_MIN=$((MACHINE_COUNT_PENDING*PD_MAINT_ADDITIONAL_NODE_MIN))
+
+        # create worker maintenance window
+        setup_maintenance_window $OCM_NAME $CD_NAMESPACE $CD_NAME $MAINT_WINDOW_MIN "$PD_MAINT_DESCRIPTION_WORKERS"
+
+        # close control plane maintenance window
+        teardown_maintenance_window $OCM_NAME $CD_NAMESPACE $CD_NAME "$PD_MAINT_DESCRIPTION_BASE"
+    fi
+
+    # log current state
+    log $OCM_NAME "upgrade" "nodes upgraded: $MACHINE_COUNT_UPGRADED/$MACHINE_COUNT_ALL"
+
     # make sure all nodes run same kubelet version
-    KUBELET_VERSION_COUNT=0
-    while [ "$KUBELET_VERSION_COUNT" != "1" ];
+    while [ ! $MACHINE_COUNT_UPGRADED -ge $MACHINE_COUNT_ALL ];
```

Comment:
```suggestion
    while [ $MACHINE_COUNT_UPGRADED -le $MACHINE_COUNT_ALL ];
```
Not greater than is less than or equal, which reads better?

Reply:
Ooh, I just noticed you had greater than or equals. My suggestion should be `lt`